language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java

```diff
@@ -0,0 +1,493 @@
+    URI uri = parseURI(params.getTextDocument().getUri());
+    analyze(uri, params.getContentChanges().get(0).getText());
+  }
+
+  @Override
+  public void didClose(DidCloseTextDocumentParams params) {
+    URI uri = parseURI(params.getTextDocument().getUri());
+    languageIdPerFileURI.remove(uri);
+    // Clear issues
+    client.publishDiagnostics(newPublishDiagnostics(uri));
+  }
+
+  @Override
+  public void didSave(DidSaveTextDocumentParams params) {
+    URI uri = parseURI(params.getTextDocument().getUri());
+    analyze(uri, params.getText());
+  }
+
+  private static URI parseURI(String uriStr) {
+    URI uri;
+    try {
+      uri = new URI(uriStr);
+    } catch (URISyntaxException e) {
+      throw new IllegalStateException(e.getMessage(), e);
+    }
+    return uri;
+  }
+
+  private void analyze(URI uri, String content) {
+    Map<URI, PublishDiagnosticsParams> files = new HashMap<>();
+    files.put(uri, newPublishDiagnostics(uri));
+    Path baseDir = workspaceDir != null ? workspaceDir : Paths.get(uri).getParent();
+    Objects.requireNonNull(baseDir);
+    Objects.requireNonNull(engine);
+    StandaloneAnalysisConfiguration configuration = new StandaloneAnalysisConfiguration(baseDir, baseDir.resolve(".sonarlint"),
+      Arrays.asList(new DefaultClientInputFile(uri, content, testFilePattern, languageIdPerFileURI.get(uri))),
+      analyzerProperties != null ? analyzerProperties : Collections.emptyMap());
+    debug("Analysis triggered on " + uri + " with configuration: \n" + configuration.toString());
+    telemetry.analysisSubmitted();
+    AnalysisResults analysisResults = engine.analyze(
+      configuration,
+      issue -> {
+        ClientInputFile inputFile = issue.getInputFile();
+        if (inputFile != null) {
+          URI uri1 = inputFile.getClientObject();
+          PublishDiagnosticsParams publish = files.computeIfAbsent(uri1, SonarLintLanguageServer::newPublishDiagnostics);
+
+          convert(issue).ifPresent(publish.getDiagnostics()::add);
+        }
+      }, logOutput);
+    // Ignore files with parsing error
+    analysisResults.failedAnalysisFiles().stream().map(ClientInputFile::getClientObject).forEach(files::remove);
+    files.values().forEach(client::publishDiagnostics);
+  }
+
+  static Optional<Diagnostic> convert(Issue issue) {
+    if (issue.getStartLine() != null) {
+      Range range = position(issue);
+      Diagnostic diagnostic = new Diagnostic();
+      DiagnosticSeverity severity = severity(issue.getSeverity());
+
+      diagnostic.setSeverity(severity);
+      diagnostic.setRange(range);
+      diagnostic.setCode(issue.getRuleKey());
+      diagnostic.setMessage(issue.getMessage() + " (" + issue.getRuleKey() + ")");
+      diagnostic.setSource(SONARLINT_SOURCE);
+
+      return Optional.of(diagnostic);
+    } else return Optional.empty();
+  }
+
+  private static DiagnosticSeverity severity(String severity) {
+    switch (severity.toUpperCase()) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define the locale to be used in this String operation. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1449)